rasa_nlu/model.py

```diff
@@ -0,0 +1,272 @@
+                updates = component.context_provides.get("process", [])
+                for u in updates:
+                    context[u] = None
+            except rasa_nlu.components.MissingArgumentError as e:
+                raise Exception("Failed to validate at component '{}'. {}".format(component.name, e.message))
+
+    def train(self, data):
+        # type: (TrainingData) -> Interpreter
+        """Trains the underlying pipeline by using the provided training data."""
+
+        self.training_data = data
+
+        context = {}
+
+        for component in self.pipeline:
+            args = rasa_nlu.components.fill_args(component.pipeline_init_args(), context, self.config)
+            updates = component.pipeline_init(*args)
+            if updates:
+                context.update(updates)
+
+        init_context = context.copy()
+
+        context["training_data"] = data
+
+        for component in self.pipeline:
+            args = rasa_nlu.components.fill_args(component.train_args(), context, self.config)
+            updates = component.train(*args)
+            if updates:
+                context.update(updates)
+
+        return Interpreter(self.pipeline, context=init_context, config=self.config)
+
+    def persist(self, path, persistor=None, create_unique_subfolder=True):
+        # type: (str, Optional[Persistor], bool) -> str
+        """Persist all components of the pipeline to the passed path. Returns the directory of the persited model."""
+
+        timestamp = datetime.datetime.now().strftime('%Y%m%d-%H%M%S')
+        metadata = {
+            "language": self.config["language"],
+            "pipeline": [component.name for component in self.pipeline],
+        }
+
+        if create_unique_subfolder:
+            dir_name = os.path.join(path, "model_" + timestamp)
+            os.makedirs(dir_name)
+        else:
+            dir_name = path
+
+        metadata.update(self.training_data.persist(dir_name))
+
+        for component in self.pipeline:
+            update = component.persist(dir_name)
+            if update:
+                metadata.update(update)
+
+        Metadata(metadata, dir_name).persist(dir_name)
+
+        if persistor is not None:
+            persistor.send_tar_to_s3(dir_name)
+        logging.info("Successfully saved model into '{}'".format(dir_name))
+        return dir_name
+
+
+class Interpreter(object):
+    """Use a trained pipeline of components to parse text messages"""
+
+    # Defines all attributes (and their default values) that will be returned by `parse`
+    default_output_attributes = {"intent": None, "entities": [], "text": ""}
+
+    @staticmethod
+    def load(meta, rasa_config):
+        # type: (Metadata, RasaNLUConfig) -> Interpreter
+        """Load a stored model and its components defined by the provided metadata."""
+        from rasa_nlu.registry import load_component_by_name
+
+        context = {"model_dir": meta.model_dir}
```

Comment:
not quite, the `meta.model_dir` is the directory the model is stored in. the configuration contains the path to the directory where ALL models are stored in (basically the parent dir). `config.update(meta.metadata)` only adds all attributes stored in the metadata file to the config. the model_dir isn't stored in the metadata though